util/messaging.py

```diff
@@ -0,0 +1,234 @@
+                    self._outbound_enqueues[id] = queue.enqueue([self._ph_content])
+                    self._outbound_closes.append(queue.close(cancel_pending_enqueues=True))
+        # response lookup for outbound RPC calls (that are waiting for a response)
+        self._waiting_for_response = {}
+        # response message id counter and lock
+        self._response_id_counter = 0
+        self._response_id_lock = Lock()
+        # local-only outbound queue for decoupling (otherwise blocking) send/receive dependencies
+        self._to_send = Queue.Queue()
+
+    def _call(self, function, args):
+        '''
+        Internal function to call a local class method as RPC function.
+        'function' - Name of the local function (member of this class)
+        'args' - Tuple of call arguments
+        '''
+        try:
+            fun = getattr(self, function, *args)
+        except AttributeError:
+            log.warn('Function not found: %s, arguments: %r' % (function, args))
+            return
+        return fun(*args)
+
+    def _send_loop(self, session, coord):
+        '''
+        Performs outbound messaging loop.
+        'session' - The current session
+        'coord' - A thread coordinator to automatically stop the loop
+        '''
+        while not coord.should_stop():
+            # get enqueue and feed_dict from local sending queue
+            enqueue, feed_dict = self._to_send.get()
+            try:
+                content = session.run(enqueue, feed_dict=feed_dict)
+            except (tf.errors.CancelledError, tf.errors.OutOfRangeError):
+                # queue got closed
+                return
+
+    def _receive(self, session, caller, is_response, response_id, function, items):
+        '''
+        Executes an inbound message. Either by calling a local RPC member function or by
+        continuing a halted/callback-ed local call that is waiting for a response.
+        'session' - The current session
+        'caller' - Id of the caller (used to respond in case of inbound RPC call)
+        'is_response' - If the message is a response (and not an inbound RPC call)
+        'response_id' - Response ID to either lookup waiting callbacks (response case) or
+            to send back a response to the caller (in case of an inbound RPC call)
+        'function' - Name of the RPC function that should be called (inbound RPC) or that
+            was called to produce this response (outbound RPC)
+        'items' - Arguments of inbound RPC call or result of outbound RPC call
+        '''
+        if is_response:
+            # we got the response for an outbound RPC call.
+            log.step('Response from: %s, function: %s, return values: %r' % (caller, function, items))
+            callback = self._waiting_for_response[response_id]
+            if callable(callback):
+                # Local response is a callable - so we call it.
+                callback(items)
```

Comment:
Is it safe like this? It feels generic, and I fear we may have exception throwing and thus breaking the codepath. Are we find we that ? Do we prefer everything to explode or handle it safely with logging ?

Reply:
Having it exploding _is_ the safe way - early and hard failing is key. I intentionally removed several try-catch blocks in this PR.